File: pkg/client/transit.go
package client

import (
	"context"
	"encoding/base64"
)

// TransitClient manages the Transit secrets engine (encryption-as-a-service).
type TransitClient struct{ c *Client }

// Transit returns a client for the Transit secrets engine.
func (c *Client) Transit() *TransitClient { return &TransitClient{c: c} }

// TransitKey describes a Transit encryption key.
type TransitKey struct {
	Name            string         `json:"name"`
	Type            string         `json:"type"`
	LatestVersion   int            `json:"latest_version"`
	MinDecryptVersion int          `json:"min_decryption_version"`
	DeletionAllowed bool           `json:"deletion_allowed"`
	Exportable      bool           `json:"exportable"`
	Keys            map[string]any `json:"keys,omitempty"`
}

// CreateKey creates a Transit encryption key. keyType defaults to "aes256-gcm96" when empty.
func (t *TransitClient) CreateKey(ctx context.Context, name, keyType string) error {
	body := map[string]string{}
	if keyType != "" {
		body["type"] = keyType
	}
	return t.c.post(ctx, "/v1/transit/keys/"+name, body, nil)
}

// GetKey reads metadata for a Transit key.
func (t *TransitClient) GetKey(ctx context.Context, name string) (*TransitKey, error) {
	var out TransitKey
	if err := t.c.get(ctx, "/v1/transit/keys/"+name, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// DeleteKey deletes a Transit key (the key must have deletion_allowed=true).
func (t *TransitClient) DeleteKey(ctx context.Context, name string) error {
	return t.c.doDelete(ctx, "/v1/transit/keys/"+name)
}

// RotateKey rotates a Transit key, incrementing its version.
func (t *TransitClient) RotateKey(ctx context.Context, name string) error {
	return t.c.post(ctx, "/v1/transit/keys/"+name+"/rotate", nil, nil)
}

// Encrypt encrypts plaintext with the named key. Returns a ciphertext string.
func (t *TransitClient) Encrypt(ctx context.Context, name string, plaintext []byte) (string, error) {
	var resp struct {
		Ciphertext string `json:"ciphertext"`
	}
	err := t.c.post(ctx, "/v1/transit/encrypt/"+name, map[string]string{
		"plaintext": base64.StdEncoding.EncodeToString(plaintext),
	}, &resp)
	return resp.Ciphertext, err
}

// Decrypt decrypts a ciphertext string with the named key. Returns plaintext bytes.
func (t *TransitClient) Decrypt(ctx context.Context, name, ciphertext string) ([]byte, error) {
	var resp struct {
		Plaintext string `json:"plaintext"`
	}
	if err := t.c.post(ctx, "/v1/transit/decrypt/"+name, map[string]string{
		"ciphertext": ciphertext,
	}, &resp); err != nil {
		return nil, err
	}
	return base64.StdEncoding.DecodeString(resp.Plaintext)
}

// Sign signs data with the named key. algorithm defaults to the key's algorithm when empty.
func (t *TransitClient) Sign(ctx context.Context, name, algorithm string, data []byte) (string, error) {
	body := map[string]string{
		"input": base64.StdEncoding.EncodeToString(data),
	}
	if algorithm != "" {
		body["signature_algorithm"] = algorithm
	}
	var resp struct {
		Signature string `json:"signature"`
	}
	err := t.c.post(ctx, "/v1/transit/sign/"+name, body, &resp)
	return resp.Signature, err
}

// Verify verifies a signature over data with the named key.
func (t *TransitClient) Verify(ctx context.Context, name, algorithm, signature string, data []byte) (bool, error) {
	body := map[string]string{
		"input":     base64.StdEncoding.EncodeToString(data),
		"signature": signature,
	}
	if algorithm != "" {
		body["signature_algorithm"] = algorithm
	}
	var resp struct {
		Valid bool `json:"valid"`
	}
	err := t.c.post(ctx, "/v1/transit/verify/"+name, body, &resp)
	return resp.Valid, err
}

// HMAC computes an HMAC over data using the named key.
func (t *TransitClient) HMAC(ctx context.Context, name, algorithm string, data []byte) (string, error) {
	body := map[string]string{
		"input": base64.StdEncoding.EncodeToString(data),
	}
	if algorithm != "" {
		body["algorithm"] = algorithm
	}
	var resp struct {
		HMAC string `json:"hmac"`
	}
	err := t.c.post(ctx, "/v1/transit/hmac/"+name, body, &resp)
	return resp.HMAC, err
}

File: web/src/shared/api/endpoints/transit.ts
import { instance } from '../axios'
import { listRequest } from '../list'

export type TransitKeyType = 'aes256-gcm96' | 'ecdsa-p256' | 'ed25519' | 'rsa-2048' | 'rsa-4096'

export interface TransitKey {
    name: string
    type: TransitKeyType
    latest_version: number
    min_decryption_version: number
    deletable: boolean
    created_at: string
    versions: Record<string, { created_at: string }>
}

export async function listTransitKeys(): Promise<string[]> {
    return listRequest('/v1/transit/keys/')
}

export async function getTransitKey(name: string): Promise<TransitKey> {
    const { data } = await instance.get<TransitKey>(`/v1/transit/keys/${name}`)
    return data
}

export async function createTransitKey(name: string, type?: TransitKeyType): Promise<TransitKey> {
    const { data } = await instance.post<TransitKey>(`/v1/transit/keys/${name}`, { type })
    return data
}

export async function deleteTransitKey(name: string): Promise<void> {
    await instance.delete(`/v1/transit/keys/${name}`)
}

export async function rotateTransitKey(name: string): Promise<void> {
    await instance.post(`/v1/transit/keys/${name}/rotate`, {})
}

export async function encrypt(name: string, plaintextB64: string): Promise<string> {
    const { data } = await instance.post<{ ciphertext: string }>(`/v1/transit/encrypt/${name}`, {
        plaintext: plaintextB64
    })
    return data.ciphertext
}

export async function decrypt(name: string, ciphertext: string): Promise<string> {
    const { data } = await instance.post<{ plaintext: string }>(`/v1/transit/decrypt/${name}`, {
        ciphertext
    })
    return data.plaintext
}

export async function sign(name: string, inputB64: string): Promise<string> {
    const { data } = await instance.post<{ signature: string }>(`/v1/transit/sign/${name}`, {
        input: inputB64,
        hash_algorithm: 'sha2-256'
    })
    return data.signature
}

export async function verify(name: string, inputB64: string, signature: string): Promise<boolean> {
    const { data } = await instance.post<{ valid: boolean }>(`/v1/transit/verify/${name}`, {
        input: inputB64,
        signature,
        hash_algorithm: 'sha2-256'
    })
    return data.valid
}

export async function hmac(name: string, inputB64: string): Promise<string> {
    const { data } = await instance.post<{ hmac: string }>(`/v1/transit/hmac/${name}`, {
        input: inputB64,
        algorithm: 'sha2-256'
    })
    return data.hmac
}
